### PR TITLE
Fix for Issue #106

### PR DESCRIPTION
**Pull request by sparkybg**
Fix for Issue 106, as described in the issue.

### CODE REVIEW

File: EF6.PG/NpgsqlMigrationSqlGenerator.cs [hunk 1]
```diff
@@ -551,6 +551,11 @@ void AppendColumn(ColumnModel column, StringBuilder sql)
                     break;
                 }
             }
+            else if (column.IsNullable != null && !column.IsNullable.Value && column.ClrDefaultValue != null)
```
Comment (roji) [inline EF6.PG/NpgsqlMigrationSqlGenerator.cs:554]: This simplification can also be done above:

```suggestion
            else if (column.IsNullable == false && column.ClrDefaultValue != null)
```

I wonder if anything is done for reference types, specifically string where an empty string can be used as the default. Any chance you can take a look at what the SQL Server provider is doing on this?

**Review by roji — APPROVED** [review]

File: EF6.PG/NpgsqlMigrationSqlGenerator.cs (hunk 1 above)
Comment (sparkybg) [reply EF6.PG/NpgsqlMigrationSqlGenerator.cs:554]: > I wonder if anything is done for reference types, specifically string where an empty string can be used as the default. Any chance you can take a look at what the SQL Server provider is doing on this?

There is no simple way to specify the default value of table column at database level in EF6, so I don't think this can be an issue at all. 

The [DefaultValue("")] data annotation does not set property's column default value on database level on any provider, including SQL Server provider.

Comment (roji) [reply EF6.PG/NpgsqlMigrationSqlGenerator.cs:554]: I'm not sure I understand... If you make an nullable int column, this PR makes it have 0 as its default value, right? Doesn't it make sense to do the same for empty string?

Comment (sparkybg) [reply EF6.PG/NpgsqlMigrationSqlGenerator.cs]: > I'm not sure I understand... If you make an nullable int column, this PR makes it have 0 as its default value, right? Doesn't it make sense to do the same for empty string?

Here's what happens when you add "int" column without any annotations:

```C#
            if (column.IsNullable != null && !column.IsNullable.Value) //This will be true
                sql.Append(" NOT NULL");

            if (column.DefaultValue != null) //this will be false
            {
                sql.Append(" DEFAULT ");
                AppendValue(column.DefaultValue, sql);
            }
            else if (!string.IsNullOrWhiteSpace(column.DefaultValueSql)) //this will be false
            {
                sql.Append(" DEFAULT ");
                sql.Append(column.DefaultValueSql);
            }
            else if (column.IsIdentity) //this will also be false
            {
                switch (column.Type)
                {
                case PrimitiveTypeKind.Guid:
                    //CreateExtension("uuid-ossp");
                    //If uuid-ossp is not enabled migrations throw exception
                    AddStatment("select * from uuid_generate_v4()");
                    sql.Append(" DEFAULT uuid_generate_v4()");
                    break;
                case PrimitiveTypeKind.Byte:
                case PrimitiveTypeKind.SByte:
                case PrimitiveTypeKind.Int16:
                case PrimitiveTypeKind.Int32:
                case PrimitiveTypeKind.Int64:
                    //TODO: Add support for setting "SERIAL"
                    break;
                }
            }
```

So we end up with "ALTER TABLE table ADD COLUMN column int4 NOT NULL, which results in error if there's already records in table, because the DB system does not know what value to set for this column on already existing records.

int? is nullable so there will not be a problem, because "NOT NULL" will not be added to the definition, and the system will set null for all present records. String is also nullable, so there will not be a problem either - same scenario.

The issue #106 only counts on non-nullable types, when there's already records in table. The resolution only adds Clr default value for non-nullable types if either "Column.DefaultValue" and "Column.DefaultValueSql" are not set. The SQL Server provider and MySQL provider are doing the same thing.

**Comment by roji** [comment]
@sparkybg how about adding a test for this?

**Comment by sparkybg** [comment]
And what should we test for?

I am "testing it" for more than 2 years on production server now. 

Here's what happens when I use postgres (with applied resolution):
`ALTER TABLE "testtable" ADD "Test" int4 NOT NULL DEFAULT 0`

SQL Server:
`ALTER TABLE [testtable] ADD [Test] [int] NOT NULL DEFAULT 0`

MYSql:
alter table testtable add column Test int not null`

MySQL also does not have "DEFAULT 0" at the end, but it still puts zeroes to the existing records. It seems that it has internal default values for at least "int" type. Postgres throws an error instead, which is what issue #106 is for.

In fact I was looking at SQL Server output for the same migration in order to figure out what should be corrected.

Here's the code from SQL Server provider for EF6:
```C#
       protected virtual void Generate(AddColumnOperation addColumnOperation)
        {
            Check.NotNull(addColumnOperation, "addColumnOperation");

            using (var writer = Writer())
            {
                writer.Write("ALTER TABLE ");
                writer.Write(Name(addColumnOperation.Table));
                writer.Write(" ADD ");

                var column = addColumnOperation.Column;

                Generate(column, writer);

                if ((column.IsNullable != null)
                    && !column.IsNullable.Value
                    && (column.DefaultValue == null)
                    && (string.IsNullOrWhiteSpace(column.DefaultValueSql))
                    && !column.IsIdentity
                    && !column.IsTimestamp
                    && !column.StoreType.EqualsIgnoreCase("rowversion")
                    && !column.StoreType.EqualsIgnoreCase("timestamp"))
                {
                    writer.Write(" DEFAULT ");

                    if (column.Type == PrimitiveTypeKind.DateTime)
                    {
                        writer.Write(Generate(DateTime.Parse("1900-01-01 00:00:00", CultureInfo.InvariantCulture)));
                    }
                    else
                    {
                        writer.Write(Generate((dynamic)column.ClrDefaultValue)); //This is what the resolution does
                    }
                }

                Statement(writer);
            }
        }
```

**Comment by sparkybg** [comment]
> @sparkybg how about adding a test for this?

Uh, sorry for misunderstanding. The tests have to be altered also.

And, two more modifications for "AlterColumnOperation" and "CreateTableOperation" along with their tests must be added also for consistency.

Is there a way to add them to this pull request or should I create a new one?

**Comment by sparkybg** [comment]
Tried, runs.

**Comment by roji** [comment]
> Is there a way to add them to this pull request or should I create a new one?

It's generally better to evolve an existing PR rather than closing and creating a new one. You can always push more commits to the PR branch - that automatically updates the PR. In some cases you'll want to "rewrite history" (e.g. rebase the PR branch on the latest version of the main branch), and in that case you need to force-push to the branch (`git push -f`). That's also fine - the PR just shows whatever's in the branch, so modify the branch in any way you want.

**Comment by sparkybg** [comment]
Yes, I figured it out, but as I said, I closed the other Pull request by mistake and didn't find a way to reopen it.

Anyway, everything considering #106 both in NpgsqlMigrationGenerator.cs and tests should be OK in the new pull request.

Sorry again - my mistake. I am working with TFS most of the time and have some bad habits. :)

**Comment by sparkybg** [comment]
Issue #106 is present again, and modifications to correct it from this pull request are lost.

Please advice what to do.

**Comment by roji** [comment]
If the changes are lost in this PR (because of force-pushing), and you don't have them locally in your own git branch, then I'm afraid there's not much we can do...
